no-undef-components failed when referencing itself in `<script setup>`

<!--
  ❗Please don't ignore this template.❗

  If you ignore it, we're just going to respond asking you to fill it out, which wastes everyone's time.
  The more relevant information you can include, the faster we can find the issue and fix it without asking you for more info.
-->

<!--
  Before posting the issue, please confirm that the problem you're getting
  is not related with your code editor configuration.
  To make sure it's not, run: yarn eslint src/your-file.vue
-->

**Checklist**

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have read the [FAQ](https://eslint.vuejs.org/user-guide/#faq) and my problem is not listed.
<!-- If you do not read the FAQ and open an issue that is listed in the FAQ, we may silently close the issue. -->

**Tell us about your environment**

- **ESLint version:** 8.28.0
- **eslint-plugin-vue version:** 9.8.0
- **Node version:** 16.17.0
- **Operating System:**  Windows 11

**Please show your full configuration:**
<!-- Paste content of your .eslintrc file -->
```js
const path = require('node:path');

module.exports = {
  env: {
    browser: true,
  },

  parser: '@babel/eslint-parser',

  parserOptions: {
    babelOptions: {
      configFile: path.resolve(__dirname, '.babelrc'),
    },
    ecmaVersion: 11,
    sourceType: 'module',
  },

  globals: {
    defineProps: 'readonly',
    defineEmits: 'readonly',
    withDefaults: 'readonly',
    VERSION: 'readonly',
  },

  ignorePatterns: ['dist/**'],

  overrides: [
    {
      files: ['*.ts'],
      extends: [
        'plugin:@typescript-eslint/eslint-recommended',
        'plugin:@typescript-eslint/recommended',
        'plugin:@typescript-eslint/recommended-requiring-type-checking',
        'plugin:prettier/recommended',
      ],

      plugins: ['@typescript-eslint'],

      parser: '@typescript-eslint/parser',

      parserOptions: {
        project: path.resolve(__dirname, './tsconfig.eslint.json'),
      },

      rules: {
        '@typescript-eslint/explicit-function-return-type': [
          'warn',
          { allowTypedFunctionExpressions: true },
        ],
        '@typescript-eslint/naming-convention': [
          'warn',
          {
            selector: 'default',
            format: ['camelCase'],
            leadingUnderscore: 'allow',
            trailingUnderscore: 'allow',
          },
          {
            selector: ['variable', 'parameter'],
            format: ['camelCase', 'PascalCase', 'UPPER_CASE'],
            leadingUnderscore: 'allow',
            trailingUnderscore: 'allow',
          },
          {
            selector: 'property',
            format: ['camelCase', 'PascalCase', 'UPPER_CASE'],
            leadingUnderscore: 'allow',
            trailingUnderscore: 'allow',
          },
          {
            selector: 'typeLike',
            format: ['PascalCase'],
          },
        ],

        '@typescript-eslint/no-explicit-any': [
          'warn',
          { ignoreRestArgs: true },
        ],

        '@typescript-eslint/no-non-null-assertion': 'off',

        '@typescript-eslint/no-unsafe-member-access': 'warn',
      },
    },
    {
      files: '*.vue',
      extends: [
        'plugin:vue/vue3-recommended',
        'plugin:@typescript-eslint/eslint-recommended',
        'plugin:@typescript-eslint/recommended',
        'plugin:prettier/recommended',
        '@vue/typescript/recommended',
      ],
      plugins: ['@typescript-eslint'],
      parser: 'vue-eslint-parser',
      parserOptions: {
        extraFileExtensions: ['.vue'],
        parser: '@typescript-eslint/parser',
        ecmaVersion: 2020,
      },
      rules: {
        '@typescript-eslint/explicit-function-return-type': [
          'warn',
          { allowTypedFunctionExpressions: true },
        ],

        '@typescript-eslint/naming-convention': [
          'warn',
          {
            selector: 'default',
            format: ['camelCase'],
            leadingUnderscore: 'allow',
            trailingUnderscore: 'allow',
          },
          {
            selector: ['variable', 'parameter'],
            format: ['camelCase', 'PascalCase', 'UPPER_CASE'],
            leadingUnderscore: 'allow',
            trailingUnderscore: 'allow',
          },
          {
            selector: 'property',
            format: ['camelCase', 'PascalCase', 'UPPER_CASE'],
            leadingUnderscore: 'allow',
            trailingUnderscore: 'allow',
          },
          {
            selector: 'typeLike',
            format: ['PascalCase'],
          },
        ],

        '@typescript-eslint/no-explicit-any': [
          'warn',
          { ignoreRestArgs: true },
        ],

        '@typescript-eslint/no-non-null-assertion': 'off',

        'vue/block-lang': ['error', { script: { lang: 'ts' } }],
        'vue/block-tag-newline': 'error',
        'vue/component-api-style': 'error',
        'vue/component-name-in-template-casing': 'error',
        'vue/component-options-name-casing': 'error',
        'vue/custom-event-name-casing': 'error',
        'vue/define-emits-declaration': 'error',
        'vue/define-macros-order': 'error',
        'vue/define-props-declaration': 'error',
        'vue/html-button-has-type': 'error',
        'vue/html-comment-content-newline': 'error',
        'vue/html-comment-content-spacing': 'error',
        'vue/html-comment-indent': 'error',
        'vue/match-component-file-name': 'error',
        'vue/match-component-import-name': 'error',
        'vue/new-line-between-multi-line-property': 'error',
        'vue/next-tick-style': 'error',
        'vue/no-boolean-default': ['warn'],
        'vue/no-duplicate-attr-inheritance': 'error',
        'vue/no-empty-component-block': 'error',
        'vue/no-multiple-objects-in-class': 'error',
        'vue/no-potential-component-option-typo': 'error',
        'vue/no-ref-object-destructure': 'error',
        'vue/no-required-prop-with-default': 'error',
        'vue/no-static-inline-styles': 'error',
        'vue/no-template-target-blank': 'error',
        'vue/no-this-in-before-route-enter': 'error',
        'vue/no-undef-components': 'error',
        'vue/no-undef-properties': ['warn'],
        'vue/no-unsupported-features': 'error',
        'vue/no-unused-properties': 'error',
        'vue/no-unused-refs': 'error',
        'vue/no-useless-mustaches': 'error',
        'vue/no-useless-v-bind': 'error',
        'vue/padding-line-between-blocks': 'error',
        'vue/padding-line-between-tags': 'error',
        'vue/prefer-prop-type-boolean-first': 'error',
        'vue/prefer-separate-static-class': 'error',
        'vue/prefer-true-attribute-shorthand': 'error',
        'vue/require-direct-export': 'error',
        'vue/require-emit-validator': ['warn'],
        'vue/require-expose': 'error',
        'vue/require-name-property': 'error',
        'vue/require-prop-comment': 'error',
        'vue/script-indent': 'error',
        'vue/static-class-names-order': 'error',
        'vue/v-for-delimiter-style': 'error',
        'vue/v-on-handler-style': 'error',
      },
    },
  ],
};

```

**What did you do?**	
<!-- Please include the actual source code causing the issue. -->
```vue
// CommentCard.vue
<template>
  <!-- ... -->
  <CommentCard />
  <!-- ... -->
</template>
<script setup lang="ts">
  // ...
</script>
```

**What did you expect to happen?**

No error 


**What actually happened?**
<!--
  Please include the actual, raw output from ESLint.
  If you are only looking at the results of your editor extension, also check the CLI results.
-->

```
/home/mister-hope/projects/waline/packages/client/src/components/CommentCard.vue
  152:9   error    The '<CommentCard>' component has been used, but not defined                                                               vue/no-undef-components
```

---

I would happy to share more if you do need this, but I do think that it's easy to make a repro, and the test file is not containing the test case where a component using `<script setup>` use itself in template.

--- a/lib/rules/no-undef-components.js
+++ b/lib/rules/no-undef-components.js
@@ -4,6 +4,7 @@
  */
 'use strict'
 
+const path = require('path')
 const utils = require('../utils')
 const casing = require('../utils/casing')
 
@@ -15,6 +16,93 @@ const casing = require('../utils/casing')
  */
 function camelize(str) {
   return str.replace(/-(\w)/g, (_, c) => (c ? c.toUpperCase() : ''))
+}
+
+class DefinedInSetupComponents {
+  constructor() {
+    /**
+     * Component names
+     * @type {Set<string>}
+     */
+    this.names = new Set()
+  }
+
+  /**
+   * @param {string[]} names
+   */
+  addName(...names) {
+    for (const name of names) {
+      this.names.add(name)
+    }
+  }
+
+  /**
+   * @see https://github.com/vuejs/core/blob/ae4b0783d78670b6e942ae2a4e3ec6efbbffa158/packages/compiler-core/src/transforms/transformElement.ts#L334
+   * @param {string} rawName
+   */
+  isDefinedComponent(rawName) {
+    if (this.names.has(rawName)) {
+      return true
+    }
+    const camelName = camelize(rawName)
+    if (this.names.has(camelName)) {
+      return true
+    }
+    const pascalName = casing.capitalize(camelName)
+    if (this.names.has(pascalName)) {
+      return true
+    }
+    // Check namespace
+    // https://github.com/vuejs/core/blob/ae4b0783d78670b6e942ae2a4e3ec6efbbffa158/packages/compiler-core/src/transforms/transformElement.ts#L305
+    const dotIndex = rawName.indexOf('.')
+    if (dotIndex > 0 && this.isDefinedComponent(rawName.slice(0, dotIndex))) {
+      return true
+    }
+    return false
+  }
+}
+
+class DefinedInOptionComponents {
+  constructor() {
+    /**
+     * Component names
+     * @type {Set<string>}
+     */
+    this.names = new Set()
+    /**
+     * Component names, transformed to kebab-case
+     * @type {Set<string>}
+     */
+    this.kebabCaseNames = new Set()
+  }
+
+  /**
+   * @param {string[]} names
+   */
+  addName(...names) {
+    for (const name of names) {
+      this.names.add(name)
+      this.kebabCaseNames.add(casing.kebabCase(name))
+    }
+  }
+
+  /**
+   * @param {string} rawName
+   */
+  isDefinedComponent(rawName) {
+    if (this.names.has(rawName)) {
+      return true
+    }
+    const kebabCaseName = casing.kebabCase(rawName)
+    if (
+      this.kebabCaseNames.has(kebabCaseName) &&
+      !casing.isPascalCase(rawName)
+    ) {
+      // Component registered as `foo-bar` cannot be used as `FooBar`
+      return true
+    }
+    return false
+  }
 }
 
 module.exports = {
@@ -109,13 +197,15 @@ module.exports = {
 
     if (utils.isScriptSetup(context)) {
       // For <script setup>
+      const definedInSetupComponents = new DefinedInSetupComponents()
+      const definedInOptionComponents = new DefinedInOptionComponents()
+
       /** @type {Set<string>} */
-      const scriptVariableNames = new Set()
       const scriptTypeOnlyNames = new Set()
       const globalScope = context.getSourceCode().scopeManager.globalScope
       if (globalScope) {
         for (const variable of globalScope.variables) {
-          scriptVariableNames.add(variable.name)
+          definedInSetupComponents.addName(variable.name)
         }
         const moduleScope = globalScope.childScopes.find(
           (scope) => scope.type === 'module'
@@ -146,39 +236,37 @@ module.exports = {
           ) {
             scriptTypeOnlyNames.add(variable.name)
           } else {
-            scriptVariableNames.add(variable.name)
+            definedInSetupComponents.addName(variable.name)
           }
         }
       }
-      /**
-       * @see https://github.com/vuejs/core/blob/ae4b0783d78670b6e942ae2a4e3ec6efbbffa158/packages/compiler-core/src/transforms/transformElement.ts#L334
-       * @param {string} name
-       */
-      const existsSetupReference = (name) => {
-        if (scriptVariableNames.has(name)) {
-          return true
+
+      // For circular references
+      const fileName = context.getFilename()
+      const selfComponentName = path.basename(fileName, path.extname(fileName))
+      definedInSetupComponents.addName(selfComponentName)
+      scriptVisitor = utils.defineVueVisitor(context, {
+        onVueObjectEnter(node, { type }) {
+          if (type !== 'export') return
+          const nameProperty = utils.findProperty(node, 'name')
+
+          if (nameProperty && utils.isStringLiteral(nameProperty.value)) {
+            const name = utils.getStringLiteralValue(nameProperty.value)
+            if (name) {
+              definedInOptionComponents.addName(name)
+            }
+          }
         }
-        const camelName = camelize(name)
-        if (scriptVariableNames.has(camelName)) {
-          return true
-        }
-        const pascalName = casing.capitalize(camelName)
-        if (scriptVariableNames.has(pascalName)) {
-          return true
-        }
-        return false
-      }
+      })
+
       verifyName = (rawName, reportNode) => {
         if (!isVerifyTargetComponent(rawName)) {
           return
         }
-        if (existsSetupReference(rawName)) {
+        if (definedInSetupComponents.isDefinedComponent(rawName)) {
           return
         }
-        // Check namespace
-        // https://github.com/vuejs/core/blob/ae4b0783d78670b6e942ae2a4e3ec6efbbffa158/packages/compiler-core/src/transforms/transformElement.ts#L305
-        const dotIndex = rawName.indexOf('.')
-        if (dotIndex > 0 && existsSetupReference(rawName.slice(0, dotIndex))) {
+        if (definedInOptionComponents.isDefinedComponent(rawName)) {
           return
         }
 
@@ -192,26 +280,10 @@ module.exports = {
       }
     } else {
       // For Options API
-
-      /**
-       * All registered components
-       * @type {string[]}
-       */
-      const registeredComponentNames = []
-      /**
-       * All registered components, transformed to kebab-case
-       * @type {string[]}
-       */
-      const registeredComponentKebabCaseNames = []
-
-      /**
-       * All registered components using kebab-case syntax
-       * @type {string[]}
-       */
-      const componentsRegisteredAsKebabCase = []
+      const definedInOptionComponents = new DefinedInOptionComponents()
 
       scriptVisitor = utils.executeOnVue(context, (obj) => {
-        registeredComponentNames.push(
+        definedInOptionComponents.addName(
           ...utils.getRegisteredComponents(obj).map(({ name }) => name)
         )
 
@@ -220,33 +292,16 @@ module.exports = {
         if (nameProperty && utils.isStringLiteral(nameProperty.value)) {
           const name = utils.getStringLiteralValue(nameProperty.value)
           if (name) {
-            registeredComponentNames.push(name)
+            definedInOptionComponents.addName(name)
           }
         }
-
-        registeredComponentKebabCaseNames.push(
-          ...registeredComponentNames.map((name) => casing.kebabCase(name))
-        )
-        componentsRegisteredAsKebabCase.push(
-          ...registeredComponentNames.filter(
-            (name) => name === casing.kebabCase(name)
-          )
-        )
       })
 
       verifyName = (rawName, reportNode) => {
         if (!isVerifyTargetComponent(rawName)) {
           return
         }
-        if (registeredComponentNames.includes(rawName)) {
-          return
-        }
-        const kebabCaseName = casing.kebabCase(rawName)
-        if (
-          registeredComponentKebabCaseNames.includes(kebabCaseName) &&
-          !casing.isPascalCase(rawName)
-        ) {
-          // Component registered as `foo-bar` cannot be used as `FooBar`
+        if (definedInOptionComponents.isDefinedComponent(rawName)) {
           return
         }
 

--- a/tests/lib/rules/no-undef-components.js
+++ b/tests/lib/rules/no-undef-components.js
@@ -23,17 +23,6 @@ tester.run('no-undef-components', rule, {
       filename: 'test.vue',
       code: `
       <script setup>
-      </script>
-
-      <template>
-        <test />
-      </template>
-      `
-    },
-    {
-      filename: 'test.vue',
-      code: `
-      <script setup>
         import Foo from './Foo.vue'
       </script>
 
@@ -118,6 +107,49 @@ tester.run('no-undef-components', rule, {
           ignorePatterns: ['custom(\\-\\w+)+']
         }
       ]
+    },
+    // circular references
+    {
+      filename: 'test.vue',
+      code: `
+      <script setup>
+      </script>
+
+      <template>
+        <test />
+      </template>
+      `
+    },
+    {
+      filename: 'FooBar.vue',
+      code: `
+      <script setup>
+      </script>
+
+      <template>
+        <FooBar />
+        <foo-bar />
+      </template>
+      `
+    },
+    {
+      filename: 'FooBar.vue',
+      code: `
+      <script>
+      export default {
+        name: 'BarFoo'
+      }
+      </script>
+      <script setup>
+      </script>
+
+      <template>
+        <FooBar />
+        <foo-bar />
+        <BarFoo />
+        <bar-foo />
+      </template>
+      `
     },
 
     // options API

